onnxruntime/core/framework/memory_info.h

```diff
@@ -0,0 +1,216 @@
+#pragma once
+#include "core/graph/basic_types.h"
+#include "core/framework/alloc_kind.h"
+#include "core/framework/data_types.h"
+#include "core/framework/execution_plan_base.h"
+#include "core/framework/mem_pattern.h"
+#include "core/framework/tensor.h"
+#include "core/framework/ort_value_name_idx_map.h"
+
+#include <iomanip>
+
+namespace onnxruntime {
+using IntervalT = std::pair<size_t, size_t>;
+using OrtValueIndex = int;
+using OrtValueName = std::string;
+
+struct MemoryInfoPerTensor {
+  MemoryBlock planned_block;
+  MemoryBlock alloced_block;
+};
+
+struct MemoryInfoMap {
+ public:
+  MemoryInfoMap() = default;
+
+  inline void AddPlannedMemory(const OrtValueIndex& idx, const MemoryBlock& mb) {
+    map_[idx].planned_block = mb;
+  }
+
+  inline void AddAllocMemory(const OrtValueIndex& idx, MemoryBlock& mb) {
+    map_[idx].alloced_block = mb;
+    if (ptr_offset == 0 || (ptr_offset > mb.offset_ && mb.offset_ != 0)) {
+      ptr_offset = mb.offset_;
+    }
+  }
+
+  inline const MemoryBlock& GetPlannedMemory(const OrtValueIndex& idx) const {
+    ORT_ENFORCE(map_.find(idx) != map_.end());
+    return map_.at(idx).planned_block;
```

Comment:
>at(idx) [](start = 16, length = 7)

With ORT_ENFORCE above at() is redundant. Perhaps a single find() result can be reused ?